lib/logstash/inputs/snmptrap.rb

```diff
@@ -47,26 +47,44 @@ def register
     end
   end # def register
 
-  public
   def run(output_queue)
     begin
       # snmp trap server
       snmptrap_listener(output_queue)
     rescue => e
       @logger.warn("SNMP Trap listener died", :exception => e, :backtrace => e.backtrace)
-      sleep(5)
-      retry
+      stoppable_call(20) { sleep(0.25) }
```

Comment:
Thats kind of what I wrote originally, but the concurrency issues that @colinsurprenant mentioned mean that the retry may occur anyway.  Oops, in this PR I am not checking for stopped in the 'start_listener' equivalent

```ruby
def run
  return if stop?
  start_listener
rescue => e
  sleep(5) unless !stop
  retry
end

# or perhaps (sorry, I could not resist)
def run
  return if stop?
  sleep(2)
  start_listener
rescue => e
  retry
end
```